### PR TITLE
[DataPipe] Reset Shuffler's iterator when NotStarted

### PR DESCRIPTION
**Pull request by ejguan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #82975 [DataLoader] Add shuffling logic for MapDataPipe
* #83202 [DataPipe] Convert MapDataPipe.shuffle to IterDataPipe
* **#83535 [DataPipe] Reset Shuffler's iterator when NotStarted**

This PR changes the behavior of `IterDataPipe` to always invoke `reset` for the state of `NotStarted`. The main reason is we normally put lazy initialization code into `reset` function. Even for the state of `NotStarted`, we should invoke `reset` to initialize those lazy variables. Otherwise, we have to manually determine if the state is `NotStarted` or `Iterating` in `__iter__` function and only manually invoke `reset` in the state of `NotStarted`.

This PR also makes `Shuffler` is able to serialize with `buffer` and `rng_state`.

The following part is removed:

~I am also add `_snapshot_state` into serialization state and during `__setstate__` only change the state to `Restored` if the original state is `Iterating`. Especially, for the case of deserializing/serializing `NotStarted` DataPipe (multiprocessing), we would invoke `set_seed` for `Shuffler`. We need the `DataPipe` remains as `NotStarted` to properly `reset`.~

I am listing all the expected behavior state transition below:
- Initial state: `NotStarted`
  - `iter` -> Call `reset` and change the state to `Iterating` 
  - serialize/deserialize -> Keep the state as `NotStarted` (will `reset` if `iter` is called afterwards)
- Initial state: `Iterating`
  - `iter` -> Call `reset` and keep the state to `Iterating` 
  - serialize/deserialize -> Change the state as `Restored`
- Initial state: `Restored`
  - `iter` -> Only change the state to `Iterating`
  - serialize/deserialize -> Not allowed

### CODE REVIEW

File: test/test_datapipe.py [hunk 1]
```diff
@@ -3145,12 +3126,15 @@ def test_simple_snapshot_graph_with_serialization(self):
...
-        self._snapshot_test_helper(graph3, expected_res=res3)
+        # TODO: shuffle is not simple graph anymore
+        graph3 = graph2
+        res3 = res2
+        #  rng = torch.Generator()
+        #  graph3 = graph2.shuffle()
+        #  rng.manual_seed(0)
+        #  torch.utils.data.graph_settings.apply_shuffle_seed(graph3, rng)
+        #  res3 = list(graph3)
+        #  self._snapshot_test_helper(graph3, expected_res=res3)
```
Comment (ejguan) [inline test/test_datapipe.py:3137]: After changing the state after `serialization`, `shuffler` becomes non-simple graph. cc: @NivekT

Comment (NivekT) [reply test/test_datapipe.py:3137]: It becomes a non-simple graph as in there is a loop in the graph?

Actually I looked into this and I don't think that is the case. See my other comment

File: torch/utils/data/datapipes/iter/combinatorics.py [hunk 2]
```diff
@@ -166,13 +173,16 @@ def __setstate__(self, state):
...
             self._number_of_samples_yielded,
-            rng_state,
+            self._snapshot_state,
         ) = state
         self._rng = random.Random()
-        self._rng.setstate(rng_state)
+        self._rng.setstate(self._rng_state)
         self._buffer = []
+        if self._snapshot_state == _SnapshotState.Iterating:
+            self._snapshot_state = _SnapshotState.Restored
```
Comment (NivekT) [inline torch/utils/data/datapipes/iter/combinatorics.py:185]: This part is causing `test_simple_snapshot_graph_with_serialization` to fail.

The reason is that `_simple_graph_snapshot_restoration` was written under the assumption that the restoration function will change the `_SnapshotState` from `Iterating` to `Restored` (rather than being done by `__setstate__`). I believe my thinking at the time was that `__setstate__` can be invoked by for reasons unrelated to snapshot restoration (e.g. serialization/multiprocessing), therefore I decided to let the restoration function change the state instead.

**Review by NivekT — APPROVED** [review]
Overall, it looks good. I took some time to debug why the unit test related to snapshotting fails, so there is one thing related to changing the `SnapshotState` that we need to discuss and confirm.

File: torch/utils/data/datapipes/iter/combinatorics.py (hunk 2 above)
Comment (NivekT) [reply torch/utils/data/datapipes/iter/combinatorics.py:185]: If you agree that `__setstate__` can be invoked for reasons unrelated for snapshotting, then we should let the restoration function handles the change in state. Let me know what you think.

File: torch/utils/data/datapipes/iter/combinatorics.py [hunk 3]
```diff
@@ -110,21 +111,23 @@ def __init__(self,
         self._enabled = True
         self._seed = None
         self._rng = random.Random()
+        # TODO: Using the initial RNG state until the full graph of DataPipe
+        #       can be serialized in the middle of iteration
+        self._rng_state = self._rng.getstate()
```
[the remark this answers is not in the document]
Comment (ejguan) [reply torch/utils/data/datapipes/iter/combinatorics.py:116]: I am saving the initial `rng_state` whenever we seeded the `rng` rather than the state in the middle of iteration. This is the same as `self._buffer` is never saved.

For proper serialization, we should save both `buffer` and `rng_state` when `__getstate__` is called.

[the remark this answers is not in the document]
Comment (NivekT) [reply torch/utils/data/datapipes/iter/combinatorics.py:116]: I see now. That makes sense. Thanks!

File: torch/utils/data/datapipes/iter/combinatorics.py (hunk 2 above)
Comment (ejguan) [reply torch/utils/data/datapipes/iter/combinatorics.py:185]: Here is the problem by removing this part of state transition:
As this PR would `reset` for the case of `NotStarted` and `Iterating` for shuffler, after deserialization, the state becomes `NotStarted` no matter serialization happens before the iteration or in the middle of iteration. Then, the DataPipe will always be `reset` once. Then, we would lose the restored `rng_state`.

File: test/test_datapipe.py [hunk 4]
```diff
@@ -1554,68 +1547,56 @@ def test_stream_reader_iterdatapipe(self):
...
+        self.assertEqual(10, len(shuffler_dp))
+        exp = list(range(100))
+
+        # Serialization Test
+        shuffler_dp = input_dp.shuffle()
+        exp = []
+        it = iter(shuffler_dp)
+        for _ in range(2):
+            exp.append(next(it))
+        shuffler_dp_copy = pickle.loads(pickle.dumps(shuffler_dp))
```
Comment (ejguan) [inline test/test_datapipe.py:1597]: Without state transition within `__setstate__`, here `shuffler_dp` has snapshot state as `Iterating` but `shuffler_dp_copy` has the snapshot state as `NotStarted`. Then `shuffler_dp_copy` would do another `reset`, which makes serialization/deserialization useless.

File: torch/utils/data/datapipes/iter/combinatorics.py (hunk 2 above)
Comment (ejguan) [reply torch/utils/data/datapipes/iter/combinatorics.py:185]: See comment: https://github.com/pytorch/pytorch/pull/83535/files#r951851003

File: test/test_datapipe.py (hunk 4 above)
Comment (NivekT) [reply test/test_datapipe.py:1597]: Is this test case what we want? You serialize/deserialize a DataPipe that has been partially read, but the expected result is that the copied DataPipe will return the full result (starting from the beginning).

When does this occur? Are DataPipes partially read prior to the start of multiprocessing?

Two somewhat related observations:
1) If `set_seed` is called prior to pickling, then the copied DP will have the seed which can be used in the next iteration as .
2) If we save the `buffer` and `rng_state (as we plan to later on), then the copied will also only have the rest of the elements (but not the first two).

**Review by NivekT — COMMENTED** [review]
I think the main issue is:
**After serializing/deserializing, we want to preserve the random state we have prior to serialization, and there are two ways to do that:**
1. Don't `reset` next time `__iter__` is called (so `rng_state` is preserved)
2. Rely on `self.seed` which is saved in serialization and will be used to set the `rng` the next time `__iter__` is called

The implementation in this PR is focused on 1. The slight complication is that it is in conflict with how the simple snapshot restoration is implemented. In particular, because `__setstate__` is changing `SnapshotState` from `Iterating` to `Restored`. The state `Restored` is also used to ensure that a DataPipe cannot be restored twice:(https://github.com/pytorch/pytorch/blob/master/torch/utils/data/datapipes/utils/snapshot.py#L33-L36).

If we want to keep going with option 1, we can either:
1. Allow DataPipes in `Restored` state to be restored in `snapshot.py`
2. Add a fourth state (maybe `Deserialized` or some other name?) that prevents `reset` but doesn't prevent restoration

File: torch/utils/data/datapipes/_hook_iterator.py [hunk 5]
```diff
@@ -8,7 +8,7 @@
 class _SnapshotState(Enum):
     r"""
     These are the snapshotting-related states that IterDataPipes can be in.
-    `NotStarted` - allows you to restore a snapshot and create an iterator without reset
+    `NotStarted` - allows you to restore a snapshot and create an iterator with reset
     `Restored` - cannot restore again, allows you to create an iterator without resetting the DataPipe
     `Iterating` - can restore, will reset if you create a new iterator
```
[the remark this answers is not in the document]
Comment (ejguan) [reply torch/utils/data/datapipes/_hook_iterator.py:13]: This PR only changes the behavior of `Shuffler`. I want to use this PR to present what is in my mind for state transition.

File: test/test_datapipe.py (hunk 4 above)
Comment (ejguan) [reply test/test_datapipe.py:1597]: I am sorry that my test case is misleading. It should be something like:
```py
shuffler_dp = input_dp.shuffle()
it = iter(shuffler_dp)
for _ in range(2):
        _ = next(it)
shuffler_dp_copy = pickle.loads(pickle.dumps(shuffler_dp))
exp = list(it)
res = list(shuffler_dp_copy)
self.assertEqual(exp, res)
```
Does this make more sense? Then, I need to serialize both `buffer` and `rng_state` to achieve so.

**Comment by NivekT** [comment]
> * Serialization/deserialization happens In the beginning of the iteration (state remains as `NotStarted`)
>   
>   * Without `set_seed`, `reset` function will be invoked when  iteration begins by both original DataPipe and copied DataPipe

Are you expecting the exact same outputs between the two DataPipes?

>  * * WIth `set_seed` called before iteration, both DataPipe should return the same value based on the random seed

Agree.

> * Serialization/deserialization happens In the middle of the iteration (state transits from `Iterating` to `Restored`)
>   
>   * No matter w/ or w/o `set_seed`, both DataPipe should return the rest of data, which means current tests is not properly implemented and I will update it.

Agree.

> TBH, I understand the current implementation prevents datapipe restored twice, which makes totally sense. But, it's still unknown to me how do we want to incorporate simple snapshot and complicate snapshot together (why restoration is required to be happened twice here). Please correct me If I am wrong. The above statement is based on my assumption that complicate snapshot relied on `__setstate__` and `__getstate__`.

I was expecting either simple or complicated snapshot restoration to be used on a DataPipe; you can't use both simultaneously on the same DataPipe. Prior to this PR, since the restoration is the one changing the SnapshotState (not during `__setstate__`, it is enforced properly.

File: test/test_datapipe.py (hunk 4 above)
Comment (NivekT) [reply test/test_datapipe.py:1597]: Yea, I think the only way this test case can pass is, as you describe, serialize/deserialize `buffer` and `rng_state`.

The alternate design would be to call some snapshot restoration functions (e.g. simple snapshot restoration).

**Comment by ejguan** [comment]
> I was expecting either simple or complicated snapshot restoration to be used on a DataPipe; you can't use both simultaneously on the same DataPipe. Prior to this PR, since the restoration is the one changing the SnapshotState (not during `__setstate__`, it is enforced properly.

I am talking about a graph of DataPipe. Let's say A -> B. If A only does simple snapshot but B does complicated snapshot and vice versa, what would the expected functional call.

**Comment by NivekT** [comment]
> > I was expecting either simple or complicated snapshot restoration to be used on a DataPipe; you can't use both simultaneously on the same DataPipe. Prior to this PR, since the restoration is the one changing the SnapshotState (not during `__setstate__`, it is enforced properly.
> 
> I am talking about a graph of DataPipe. Let's say A -> B. If A only does simple snapshot but B does complicated snapshot and vice versa, what would the expected state of each DataPipe and how to invoke the snapshot call?

If A does simple snapshot and B does complicated (i.e. restore buffer and rng), that is fine.

If B does simple snapshot, then nothing needs to be done for A since B is dependent on A.

**Comment by ejguan** [comment]
> If A does simple snapshot and B does complicated (i.e. restore buffer and rng), that is fine.
> 
> If B does simple snapshot, then nothing needs to be done for A since B is dependent on A, and in the process of fast-forwarding B, A (assuming it starts from a clean state) will be properly restored.

Serialization might work. But, for deserialization, it's more complicated IMHO. If prior DataPipe is complicated snapshotted, when deserialized, its state is changed to `Restored` and all related attributes are updated to the new state. But, the end DataPipe needs to fast forward a certain iteration to achieve the same state, which will also affect the prior DataPipe. And, only after fast forwarded, is the end DataPipe changed to `Restored`.

**Comment by NivekT** [comment]
> Serialization might work. But, for deserialization, it's more complicated IMHO. If prior DataPipe is complicated snapshotted, when deserialized, its state is changed to `Restored` and all related attributes are updated to the new state. But, the end DataPipe needs to fast forward a certain iteration to achieve the same state, which will also affect the prior DataPipe. And, only after fast forwarded, is the end DataPipe changed to `Restored`.

I think this part is why I decided against changing its state to `Restored` in `__setstate__` in my PR and thusfar we haven't serialized the buffer. If we restore buffer and etc in `__setstate__` and use simple snapshot restoration on a DataPipe downstream, it will have to forcibly reset all the DataPipe states before it in order for simple snapshot restoration to work.

Is there a specific use case where users want the buffer stored in serialization/deserialization rather than using the snapshotting API that we will provide?

**Comment by ejguan** [comment]
> I think this part is why I decided against changing its state to `Restored` in `__setstate__` in my PR and thusfar we haven't serialized the buffer.
> 
> If we restore buffer and etc in `__setstate__` and use simple snapshot restoration on a DataPipe downstream, it will have to forcibly reset all the DataPipe states before it in order for simple snapshot restoration to work.

Then, how you are going to figure out which `DataPipe` is restored by the complicated restoration when restoring the downstream simple snapshotted DataPipe? And, (different topic), it seems [the current implementation](https://github.com/pytorch/pytorch/blob/58170fb8aafc955258609693cd368cf8ecc8ff2b/torch/utils/data/datapipes/utils/snapshot.py#L41) for simple restoration only `reset` the last `DataPipe`.

> Is there a specific use case where users want the buffer stored in serialization/deserialization rather than using the snapshotting API that we will provide?

I don't expect users using any thing other than snapshotting API. For ourselves, we might need it for future proof (elastic multiprocessing), but it's not needed.

TBH, I can modify the serialization test from my PR, which uses `pickle.dumps` and `pickle.loads` to make sure all tests passes even without the change of this state transition. See: https://github.com/pytorch/pytorch/pull/83535/files#r951851003
Basically, I will manually change the state after `pickle.loads` to make it work.

File: test/test_datapipe.py (hunk 1 above)
Comment (ejguan) [reply test/test_datapipe.py]: I am changing this PR without touching the state of `Shuffler`. So, `shuffler` is still working properly with simple snapshot by always `reset`.

File: torch/utils/data/datapipes/iter/combinatorics.py (hunk 2 above)
Comment (ejguan) [reply torch/utils/data/datapipes/iter/combinatorics.py]: This part is removed.

File: test/test_datapipe.py (hunk 4 above)
Comment (ejguan) [reply test/test_datapipe.py]: Revise the test case

File: test/test_datapipe.py [hunk 6]
```diff
@@ -1554,68 +1547,65 @@ def test_stream_reader_iterdatapipe(self):
...
+            shuffler_dp = input_dp.shuffle(buffer_size=buffer_size)
+            it = iter(shuffler_dp)
+            for _ in range(2):
+                next(it)
+            shuffler_dp_copy = pickle.loads(pickle.dumps(shuffler_dp))
+            _simple_graph_snapshot_restoration(shuffler_dp_copy.datapipe, shuffler_dp.datapipe._number_of_samples_yielded)
+
+            exp = list(it)
+            shuffler_dp_copy._snapshot_state = _SnapshotState.Restored
+            self.assertEqual(exp, list(shuffler_dp_copy), f"Failed Shuffler serialization test with {buffer_size=}")
```
Comment (ejguan) [inline test/test_datapipe.py:1604]: Changed the test case:
- Use `_simple_graph_snapshot_restoration` the input DataPipe
- Use complicated restoration for `Shuffler`
cc: @NivekT

File: torch/utils/data/datapipes/iter/combinatorics.py [hunk 7]
```diff
@@ -145,6 +144,8 @@ def reset(self) -> None:
         self._buffer = []
         if self._enabled and self._seed is None:
             self._seed = int(torch.empty((), dtype=torch.int64).random_().item())
+        self._rng.seed(self._seed)
+        self._seed = None
```
Comment (rgtjf) [inline torch/utils/data/datapipes/iter/combinatorics.py:148]: Thanks for the great PR! 

I'm wondering why we need to set self._seed to None here. I encountered a problem where iterating twice caused the dataset to lose determinism.

```python
dp1 = IterableWrapper(range(10)).shuffle().set_seed(1) 
dp2 = IterableWrapper(range(10)).shuffle().set_seed(1)

# First iteration
all_match1 = all(x == y for x, y in zip(dp1, dp2))  # true
# Second iteration
all_match2 = all(x == y for x, y in zip(dp1, dp2))  # false
```
I would appreciate any clarification you can provide!
